### PR TITLE
Implements a disguise decoration to Spies. 

### PR DESCRIPTION
**Pull request by GraionDilach**
Deprecates WithDecorationDisguised in favor of a refactor in Disguise.
spypip
Opened eye when disguised (animated, thanks @x-a-n-a-x ), closed eye when undisguised.

Implements #7466.

### CODE REVIEW

**Comment by Phrohdoh** [comment]
There's no reason to use `pipdis` as the filename, we aren't limited in character count.

File: OpenRA.Mods.RA/Traits/Disguise.cs [hunk 1]
```diff
@@ -108,6 +124,9 @@ public Color RadarColorOverride(Actor self)
 
 		void DisguiseAs(Actor self, Actor target)
 		{
+			var um = self.TraitOrDefault<UpgradeManager>();
```
Comment (Phrohdoh) [inline OpenRA.Mods.RA/Traits/Disguise.cs]: Use `Requires<>` and assign this in the ctor.

File: OpenRA.Mods.RA/Traits/Disguise.cs [hunk 2]
```diff
@@ -128,16 +147,33 @@ void DisguiseAs(Actor self, Actor target)
...
 				AsTooltipInfo = null;
 				AsPlayer = null;
 				AsSprite = null;
+
 			}
 
 			foreach (var t in self.TraitsImplementing<INotifyEffectiveOwnerChanged>())
 				t.OnEffectiveOwnerChanged(self, oldEffectiveOwner, AsPlayer);
+
+			if ((Disguised != oldDisguiseSetting) && (um != null))
```
Comment (Phrohdoh) [inline OpenRA.Mods.RA/Traits/Disguise.cs]: Unnecessary parenthesis pairs.

**Comment by GraionDilach** [comment]
Will fix these - and Travis - on Monday, sorry.

File: OpenRA.Mods.RA/Traits/Disguise.cs (hunk 1 above)
Comment (ScottNZ) [inline OpenRA.Mods.RA/Traits/Disguise.cs]: `Lazy<T>` would be more suitable if UpgradeManager is actually optional

**Comment by epicelite** [comment]
How about no eye at all when undisguised?

**Comment by Phrohdoh** [comment]
That was the original plan.

**Comment by chrisforbes** [comment]
That seems like a better plan. The open/closed eyes are hard to distinguish.

**Comment by GraionDilach** [comment]
Concerns addressed - renamed pipdis to pip-disguise, removed closed eye, fixed StyleCop and code issues.

File: mods/ra/rules/infantry.yaml [hunk 3]
```diff
@@ -214,11 +214,21 @@ SPY:
...
+		Offset: 4, -2
+		ZOffset: 256
+		UpgradeTypes: disguise
+		UpgradeMinEnabledLevel: 1
 	Armament:
 		Weapon: SilencedPPK
 	AttackFrontal:
 	Voiced:
 		VoiceSet: SpyVoice
+
```
Comment (Mailaender) [inline mods/ra/rules/infantry.yaml]: Spurious newline.

**Comment by Mailaender** [comment]
Don't forget to add yourself to https://github.com/OpenRA/OpenRA/blob/bleed/AUTHORS

File: mods/ra/rules/infantry.yaml (hunk 3 above)
Comment (GraionDilach) [inline mods/ra/rules/infantry.yaml]: Ah, yeah, thanks.

**Comment by GraionDilach** [comment]
I believe in the rule of three - and this is only the second one.

File: OpenRA.Mods.RA/Traits/Disguise.cs [hunk 4]
```diff
@@ -138,8 +158,22 @@ void DisguiseAs(Actor self, Actor target)
...
 				t.OnEffectiveOwnerChanged(self, oldEffectiveOwner, AsPlayer);
+
+			if (Disguised != oldDisguiseSetting && um.Value != null)
+			{
+				foreach (var u in info.Upgrades)
+				{
+					if (!um.Value.AcceptsUpgrade(self, u))
+						continue;
+
+					if (Disguised)
```
Comment (Mailaender) [inline OpenRA.Mods.RA/Traits/Disguise.cs]: Currently the revoking does not work. 
```csharp
if (Disguised && AsPlayer != self.Owner)
```

might fix it.

Comment (GraionDilach) [inline OpenRA.Mods.RA/Traits/Disguise.cs]: How did you tested this out? I honestly couldn't even figure out an ingame way to drop the disguise...

Comment (Mailaender) [inline OpenRA.Mods.RA/Traits/Disguise.cs]: Right-click on the spy again or shoot an enemy (or force fire on the ground).

Comment (GraionDilach) [inline OpenRA.Mods.RA/Traits/Disguise.cs]: Thanks.

**Comment by reaperrr** [comment]
Needs rebase.

File: OpenRA.Mods.RA/Traits/Disguise.cs (hunk 4 above)
Comment (GraionDilach) [inline OpenRA.Mods.RA/Traits/Disguise.cs]: Turned out that's not the issue. Due to UpgradeMaxAcceptsLevel defaults to 1, the AcceptsUpgrade check notices the Spy cannot taker another Disguise upgrade and returns false even before this line.

**Comment by GraionDilach** [comment]
Rebased and fixed the issue about the eye never disappearing, see my last comment upon the outdated diff/second commit's message.

**Comment by GraionDilach** [comment]
BTW, does this need an upgrade rule?

File: OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs [hunk 5]
```diff
@@ -125,6 +125,14 @@ public void RevokeUpgrade(Actor self, string upgrade, object source)
 			NotifyUpgradeLevelChanged(s.Traits, self, upgrade, -1);
 		}
 
+		//Returns true if the actor can work with the upgrade. Does not check the actual level of the upgrade.
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs]: What does "can work with the upgrade" mean?
Also these comments should be `<summary>`.

File: OpenRA.Mods.RA/Traits/Disguise.cs [hunk 6]
```diff
@@ -61,18 +62,14 @@ class DisguiseInfo : ITraitInfo
 	{
 		[VoiceReference] public readonly string Voice = "Action";
 
+		[Desc("Grants upgrades when disguised.")]
```
Comment (Phrohdoh) [inline OpenRA.Mods.RA/Traits/Disguise.cs]: Suggested English: `Upgrades to grant when disguised.`

**Comment by penev92** [comment]
The trait has been released for a week now, and was not even used in the default mods. I'd say no need.

File: OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs (hunk 5 above)
Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs]: That the upgrade is defined on the actor.

I didn't see many such descriptions in summary, sorry about that.

Comment (penev92) [inline OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs]: "uses the given upgrade" ?

File: OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs [hunk 7]
```diff
@@ -125,6 +125,14 @@ public void RevokeUpgrade(Actor self, string upgrade, object source)
...
+		public bool AcknowledgesUpgrade(Actor self, string upgrade)
+		{
+			UpgradeState s;
+			return upgrades.Value.TryGetValue(upgrade, out s);
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs]: Looks like this can be simplified to `return upgrades.Value.ContainsKey(upgrade);`

**Comment by GraionDilach** [comment]
Tomorrow I'll fix the issues. >_>

**Comment by GraionDilach** [comment]
Resolved all issues aforementioned.

**Comment by obrakmann** [comment]
Code looks good and works as advertised. Not sure if I like the art, though. It's very hard to make out unless your're zoomed in in my opinion, and the eye moves too fast for my liking. But I'll go with the majority's opinion here. :+1:

**Comment by chrisforbes** [comment]
:+1:

**Comment by GraionDilach** [comment]
I can give another try at art, I don't mind that.

**Comment by GraionDilach** [comment]
Remade art, now it's a bit bigger, has an outer edge and slower (both because of smoother movement and Tick increased from 250 to 300).

I really hope it's okay now.

newopenraspy

File: OpenRA.Mods.RA/Traits/Disguise.cs [hunk 8]
```diff
@@ -61,18 +62,14 @@ class DisguiseInfo : ITraitInfo
 	{
 		[VoiceReference] public readonly string Voice = "Action";
 
+		[Desc("Upgrades to grant when disguised.")]
+		public readonly string[] Upgrades = { "disguise" };
```
Comment (Mailaender) [inline OpenRA.Mods.RA/Traits/Disguise.cs]: This needs a `[UpgradeGrantedReference]` now. See https://github.com/OpenRA/OpenRA/pull/8321/

**Comment by Mailaender** [comment]
> Remade art, now it's a bit bigger, has an outer edge and slower (both because of smoother movement and Tick increased from 250 to 300).

Looks much better now! :eyes:

**Comment by GraionDilach** [comment]
Rebased, added reference and the same asset to the TS Chameleon Spy as being an additional Disguise trait user.

**Comment by Mailaender** [comment]
I guess we can live without an upgrade rule for the `WithDecorationDisguised` as it was never used.

**Comment by Mailaender** [comment]
Works as promised. :white_check_mark:

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/78002d14d8191d3f42e701733a5bb0d0a9c51467...a084fa5b429f447257ee5e77df3a6f6d44e757be)

**Comment by Mailaender** [comment]
Well done!

**Comment by Phrohdoh** [comment]
@GraionDilach Nice job on this!

**Pull request merged**